src/runtime/minrpc/minrpc_logger.h

```diff
@@ -0,0 +1,266 @@
+        break;
+      }
+      case kTVMNullptr:
+        this->LogString("Nullptr");
+        break;
+      case kTVMStr: {
+        this->LogString("\"");
+        this->LogString(value.v_str);
+        this->LogString("\"");
+        break;
+      }
+      case kTVMBytes: {
+        TVMByteArray* bytes = static_cast<TVMByteArray*>(value.v_handle);
+        int len = bytes->size;
+        this->LogVal<int64_t>("(Bytes) [size]: ", len);
+        if (PRINT_BYTES) {
+          this->LogString(", [Values]:");
+          this->LogString(" { ");
+          if (len > 0) {
+            this->LogVal<uint64_t>("", (uint8_t)bytes->data[0]);
+          }
+          for (int j = 1; j < len; j++) this->LogVal<uint64_t>(" - ", (uint8_t)bytes->data[j]);
+          this->LogString(" } ");
+        }
+        break;
+      }
+      default: {
+        this->LogString("ERROR-kUnknownTypeCode)");
+        break;
+      }
+    }
+    this->LogString("; ");
+  }
+
+  void LogString(const char* s) { os_ << s; }
+
+  void LogStr(std::string s) { os_ << s; }
+
+  void LogHandleName(std::string name) {
+    if (name.length() > 0) {
+      os_ << " <" << name.c_str() << ">";
+    }
+  }
+
+  template <typename T>
+  void LogVal(const char* s, T val) {
+    os_ << s << val;
+  }
+
+  void LogDLDevice(const char* s, DLDevice* dev) {
+    os_ << s << "(" << dev->device_type << "," << dev->device_id << ")";
+  }
+
+  void LogDLData(const char* s, DLDataType* data) {
+    os_ << s << "(" << (uint16_t)data->code << "," << (uint16_t)data->bits << "," << data->lanes
+        << ")";
+  }
+
+  std::stringstream LogTime();
+
+  void OutputLog();
+
+ private:
+  std::stringstream os_;
+};
+
+/*!
+ * \brief A wrapper for a MinRPCReturns object, that also logs the responses.
+ *
+ * \tparam ReturnInterface* underlying MinRPCReturns that generates the responses.
+ */
+class MinRPCReturnsWithLog : public ReturnInterface {
+ public:
+  /*!
+   * \brief Constructor.
+   * \param io The IO handler.
+   */
+  explicit MinRPCReturnsWithLog(ReturnInterface* next);
+  ~MinRPCReturnsWithLog();
+  void ReturnVoid();
+  void ReturnHandle(void* handle);
+  void ReturnException(const char* msg);
+  void ReturnPackedSeq(const TVMValue* arg_values, const int* type_codes, int num_args);
+  void ReturnCopyAck(uint64_t* num_bytes, uint8_t** data_ptr);
+  void ReturnLastTVMError();
+  void ThrowError(RPCServerStatus code, RPCCode info);
+  void processValues(const TVMValue** values, const int** tcodes, int* num_args);
+  void resetCurrHandleName(RPCCode code);
+  void updateCurrHandleName(const char* name);
+  void getHandleName(void* handle);
+  void releaseHandleName(void* handle);
+  Logger* getLogger();
+
+ private:
+  void registerHandleName(void* handle);
+  ReturnInterface* next_;
+  std::string CurrHandleName;
+  std::unordered_map<void*, std::string> array_tracker;
+  RPCCode code;
+  Logger logger_;
+  Logger* logger = &logger_;
```

Comment:
Is this needed? I feel like it adds a bit of confusion when reading implementation.

Reply:
thanks, fixed it!